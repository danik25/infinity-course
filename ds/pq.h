#ifndef _PRIORITY_QUEUE_
#define _PRIORITY_QUEUE_

#include <stddef.h> /* size_t */



/****************************************************************
*                <<<<<<<< ASSIGNMENT NAME >>>>>>>>              *
* Related files: Priority.c, Priority_test.c                    *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Or Stopiky                                          *
* Last update: 17/04/19                                         *
****************************************************************/  
/* API DESCRIPTION:
 *
 * Priority queues are a type of container adaptors, specifically designed such 
 * that its first element is always the greatest of the elements it contains, 
 * according to some strict ordering criterion supplied by the user.
 *
 *
 * 
 */
 

typedef struct pqueue pqueue_t;


/* Description: 
            Creates an empty priority queue.
 *                  
 * Params: 
 *      Is_before - a user supplied function that will determine the 
 *      priority for sorting the elements of the queue.
 *      Param - an additional paramater for the use of the user.      
 *        
 * Return Value: 
 *        Pointer to the newly created queue.        
*/
pqueue_t *PQCreate(int (is_before)(void *data1, void *data2, void *param),
                                                                  void *param);


/* Description:
 *          Free all allocated memory of the queue and destroy the queue. 
 *
 * Params:
 *      queue - the priority queue to be destroyed.  
 *        
 *        
 * Return Value: void       
*/
void PQDestroy(pqueue_t *queue);



/* Description: 
 *          Unserts an element to the queue in a location determined by its
 *          priority.
 *
 * Params: 
 *       queue - the queue to insert to.    
 *       data - the data to insert. 
 *        
 * Return Value: 
 *         status - success(0) or failure(1).
*/
int PQEnqueue(pqueue_t *queue, void *data);



/* Description:
 *          Removes the element with the highest priority from the queue.
 *          Time complexity: O(n)
 *
 * Params:
 *     queue - the queue to remove from.
 *               
 * Return Value: 
 * 
 *      The data of the removed element.
 *      If the queue is empty - returns NULL.     
*/
void *PQDequeue(pqueue_t *queue);



/* Description:
 *          Shows the data of the element with the highest priority.
 *
 * Params:
 *      queue - the queue to look at.  
 *             
 * Return Value: 
 *          The data of the highest priority element.
 *          If the queue is empty - returns NULL.
*/
void *PQPeek(const pqueue_t *queue);



/* Description:
 *          Checks whether the queue is empty or not.
 *
 * Params:
 *      queue - the queue to look in.
 *            
 * Return Value: 
 *      Boolean (TRUE(1) / FALSE(0)).           
*/
int PQIsEmpty(const pqueue_t *queue);



/* Description:
 *          Counts number of element in the queue.
 *
 * Params:
 *     queue - the queue to count.   
 *        
 * Return Value: 
 *      Number of elements.
 *          
*/
size_t PQSize(const pqueue_t *queue);



/* Description:
 *          Free all allocated memory of the queue's elements leaving an
 *          empty queue without destroying it.
 *
 * Params:
 *      queue - te queue to clear.
 *        
 *        
 * Return Value: 
 *              void.
*/
void PQClear(pqueue_t *queue);



/* Description:
 *          Removes a specific element from the queue that matches a user 
 *          specified data.
 * 
 * Params:
 *     is_match_func - used to compare the user data with the queue elements and
 *                     find if a match exist (boolean function).    
 *                     data - the user supplied data for comarison.
 *        
 * Return Value: 
 *              The data of the removed element or NULL if element was not 
 *              found.
*/
void *PQRemove(pqueue_t *queue, int (is_match_func)(void *data, void *args),
                                                                    void *args);
                                                                    
#endif /* _PRIORITY_QUEUE_ */