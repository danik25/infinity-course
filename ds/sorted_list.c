/****************************************************************
*                Data Structures - Sorted List                  *
* Related files:    sorted_list.c    dl_list.h                  *
*	            sorted_list.h    sorted_list_test.c	            *
* #Version: V 1.4                                               *
* Writer: Dani Benjamin                                	        *
* Reviewer: Tom Valershtein                                    	*
* Last update: 16.4.19                                          *
****************************************************************/
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/ 

#include "sorted_list.h"
#include "dl_list.h"


struct sorted_list {

    sort_compare_function_t *IsBefore;
    void * param;
    dll_t *list;

};


sorted_list_t *SortCreate(sort_compare_function_t is_before, void *param)
{
    sorted_list_t *sorted = (sorted_list_t*) malloc(sizeof(sorted_list_t));
    dll_t *dll = NULL; 
    
    if(!sorted)
    {
        return (NULL);
    }
    
    dll = DLCreate();

    if(!dll)
    {
        free(sorted);
        return (NULL);
    }
    
    sorted->IsBefore = is_before;
    sorted->list = dll;
    sorted->param = param;
    
    return (sorted);
}

void SortDestroy(sorted_list_t *sorted_list)
{
    assert(sorted_list);
    
    DLDestroy(sorted_list->list);
    free(sorted_list);
    sorted_list = NULL;
}

int SortIsEmpty(const sorted_list_t *sorted_list)
{
    assert(sorted_list);
    
    return (DLIsEmpty(sorted_list->list));
}

size_t SortSize(const sorted_list_t *sorted_list)
{
    assert(sorted_list);
    
    return (DLCount(sorted_list->list));
}

sorted_iter_t SortEnd(const sorted_list_t *sorted_list)
{
    assert(sorted_list);
    
    return (DLEnd(sorted_list->list));
}

sorted_iter_t SortBegin(const sorted_list_t *sorted_list)
{
    assert(sorted_list);
    
    return (DLBegin(sorted_list->list));
}

sorted_iter_t SortNext(const sorted_iter_t location)
{
    assert(location);
    
    return (DLNext(location));
}

sorted_iter_t SortPrev(const sorted_iter_t location)
{
    assert(location);
    
    return (DLPrev(location));
}

void *SortGetData(const sorted_iter_t location)
{
    assert(location);
    
    return (DLGetData(location));
}

int SortIsSameIter(const sorted_iter_t first_iter, const sorted_iter_t second_iter)
{
    assert(first_iter);
    assert(second_iter);
    
    return (DLSameIter(first_iter, second_iter));
}

sorted_iter_t SortErase(sorted_iter_t location)
{
    assert(location);
    
    return (DLErase(location));
}

sorted_iter_t SortInsert(sorted_list_t *sorted_list, void *data_to_insert)
{
    sorted_iter_t inserted_node = NULL;
    sorted_iter_t answer = NULL; 
    
    assert(sorted_list);
    
    inserted_node = SortBegin(sorted_list);
    while(inserted_node != SortEnd(sorted_list) && 
         (!sorted_list->IsBefore(data_to_insert, SortGetData(inserted_node), 
                                                    sorted_list->param)))
    {
        inserted_node = SortNext(inserted_node);
    }
    answer = (DLInsert(sorted_list->list, inserted_node, data_to_insert));
    if(answer == SortEnd(sorted_list))
    {
        return (NULL);
    }
    return (answer);
}

void *SortPopFront(sorted_list_t *sorted_list)
{
    void * data = NULL;
    
    assert(sorted_list);
    
    if(SortIsEmpty(sorted_list))
    {
        return (NULL);
    }
    
    data = SortGetData(SortBegin(sorted_list));
    DLPopFront(sorted_list->list);
    return (data);
    
}

void *SortPopBack(sorted_list_t *sorted_list)
{
    void * data = NULL;
    
    assert(sorted_list);
    
    if(SortIsEmpty(sorted_list))
    {
        return (NULL);
    }
    
    data = SortGetData(SortPrev(SortEnd(sorted_list)));
    DLPopBack(sorted_list->list);
    return (data);
    
}

int SortForEach(sorted_iter_t from, sorted_iter_t to,
						 	         sort_action_function_t func , void *param)
{
    assert(from);
    assert(to);
    assert(func);
    
    return (DLForEach(from, to, func , param));
}

sorted_iter_t SortFind(sorted_list_t *sorted_list, sorted_iter_t from, 
                  sorted_iter_t to, sort_is_match_function_t func, void *param)
{
    assert(sorted_list);
    assert(from);
    assert(to);
    assert(func);
    
    return (DLFind(sorted_list->list, from, to, func, param));
}

void SortMerge(sorted_list_t *to, sorted_list_t *from)
{
    sorted_iter_t node_to = NULL;
    sorted_iter_t node_from = NULL;
    sorted_iter_t node_from_end = NULL;
    
    assert(to);
    assert(from);
    
    if(SortIsEmpty(from))
    {
        free(from->list);
        free(from);
        from = NULL;
        return;
    }
    
    node_to = SortBegin(to);
    node_from = SortBegin(from);
    node_from_end = node_from;
    
    for(node_to = SortBegin(to); node_to != SortEnd(to) && 
            node_from != SortEnd(from); node_to = SortNext(node_to))
    {
        if(to->IsBefore(SortGetData(node_from), 
                                SortGetData(node_to), to->param))
        {
            while(node_from_end != SortEnd(from) && 
                        to->IsBefore(SortGetData(node_from_end), 
                                SortGetData(node_to), to->param))
            {
                node_from_end = SortNext(node_from_end);
            }
            DLSplice(from->list, node_from, node_from_end, node_to);
            node_from = node_from_end;                    
        }
    }
    while(node_from_end != SortEnd(from))
    {
        node_from_end = SortNext(node_from_end);
    }
    DLSplice(from->list, node_from, node_from_end, node_to);
    
    free(from->list);
    free(from);
    from = NULL;
    
}




