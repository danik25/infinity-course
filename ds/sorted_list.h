/****************************************************************
*                Data Structures - Sorted List                  *
* Related files:    sorted_list.c    dl_list.h                  *
*	            sorted_list.h    sorted_list_test.c	            *
* #Version: V 1.4                                               *
* Writer: Dani Benjamin                                	        *
* Reviewer: Tom Valershtein                                    	*
* Last update: 16.4.19                                          *
****************************************************************/ 
/*
 * Sorted list is a list that its elements are ordered 
 * by a function provided by user. 
 * The list can be traversed in both forward and backward direction.
*/  

/* API Design:
 *	
 *      This API is desgined to store elements in a sorted way, in a list.
 *		This design supports adding and removing elements. 
 *		Note that the sorting is done by the user provided compare function. 
 *		This design supports any data type. 
*/


#ifndef _SORTED_LIST_
#define _SORTED_LIST_

#include <stddef.h> /* size_t */


typedef int (sort_action_function_t)(void *data, void *param);
typedef int (sort_is_match_function_t)(void *data, void *param);

/* Compare between two data.
 *
 * Params:
 *		data1 - data from new element.
 *		data2 - data from list's elements.
 *      param - additional parameter provided by user.
 *
 * Return Value: 
 *      boolean (TRUE(1) / FALSE(0)).
*/
typedef int (sort_compare_function_t)(void *data1, void *data2, void *param);

typedef struct sorted_list sorted_list_t;
typedef void * sorted_iter_t;


/* Create an empty sorted list, 
 * actually containing two dummy nodes.
 *	
 * Param:
 *		func - compare function.
 *      param - additional parameter provided by user.
 *	  
 * Return Value: 
 *		  pointer to the newly created list. 
*/
sorted_list_t *SortCreate(sort_compare_function_t is_before, void *param);


/* Destroy the sorted list - 
 * free all allocated memory from a given position.
 *
 * Param: 
 *      sorted_list - the sorted list to work on.
 *
 * Time complexity: O(n).
*/
void SortDestroy(sorted_list_t *sorted_list);	  


/* Checks whether or not the sorted list is empty.
 * 
 * Param: 
 *      sorted_list - the sorted list to work on.
 *
 * Return Value: 
 *      boolean (TRUE(1) / FALSE(0)).
*/
int SortIsEmpty(const sorted_list_t *sorted_list);


/* Counts elements in a given sorted list.
 *  
 * Param: 
 *      sorted_list - the sorted list to work on.
 *
 * Return value: 
 *      number of elements in the sorted list.
*/
size_t SortSize(const sorted_list_t *sorted_list);


/* Finds the last element of a given sorted list. 
 * 
 * Param: 
 *      sorted_list - the sorted list to work on.	
 *	  
 * Return value: 
 *      a pointer to the end of the sorted list (tail dummy).
*/
sorted_iter_t SortEnd(const sorted_list_t *sorted_list);


/* Finds the first element of a given sorted list. 
 *	 
 * Param: 
 *      sorted_list - the sorted list to work on.
 *	  
 * Return value: 
 *      a pointer to the beginning of the sorted list.
 *      in case of empty list return dummy tail.
*/
sorted_iter_t SortBegin(const sorted_list_t *sorted_list);


/* Gives the next element from a given location. 
 *
 * Param: 
 *      location - specific node to check.
 *		  
 * Return value: 
 *      a pointer to the next element of a given location.
 *      In case location points to tail dummy, 
 *      it is undefined behavier- return NULL. 
*/
sorted_iter_t SortNext(const sorted_iter_t location);


/* Gives the previous element from a given location. 
 *	
 * Param:  
 *      location - specific node to check.
 *                 
 * Return value: 
 *      a pointer to the prev element of a given location.
 *      in case location points to first element or in case of empty list, 
 *      it is undefined behavier- return NULL. 
*/
sorted_iter_t SortPrev(const sorted_iter_t location);


/* Extracts data from a given element.
 *
 * param: 
 *      location - specific node to check.
 *
 * Return value: 
 *      the data from the given location. 
*/
void *SortGetData(const sorted_iter_t location);


/* Compares addresses of two given iterators.
 *
 * Params: 
 *      first_iter - first pointer to a node. 
 *      second_iter - second pointer to a node.
 *
 * Return Value: 
 *      boolean (TRUE(1) / FALSE(0)).
*/
int SortIsSameIter(const sorted_iter_t first_iter, 
                                               const sorted_iter_t second_iter);


/* Erase an element in a specific location.
 *
 * Params:
 *      location - pointer to the node to be erased.
 *	
 * Return value : 
 *      pointer to next node after the erased node.
 *      in case location points to end, 
 *      it is undefined behavier- return NULL. 
*/
sorted_iter_t SortErase(sorted_iter_t location);


/* Creates and inserts an element to the given sorted list,
 * so the list will stay sorted (insertion location determined by 
 * the compare function).
 *
 * Params:
 *      sorted_list - a sorted list to insert an element to.	 
 *      data_to_insert - the data to insert.
 *	
 * Return value : 
 *      pointer to inserted node if succeded. 
 *      NULL if failed.  
 *
 * Time complexity: O(n).
*/
sorted_iter_t SortInsert(sorted_list_t *sorted_list, void *data_to_insert);


/* Remove a node from the beginning of a sorted list.
 *
 * Param: 
 *      sorted_list - a sorted list to remove an element from.
 *
 * Return value : 
 *      pointer to the data of the removed node, if succeded.
 *      NULL if failed. 
*/
void *SortPopFront(sorted_list_t *sorted_list);


/* Remove a node from the end of a sorted list.
 *
 * Param: 
 *      sorted_list - a sorted list to remove an element from.
 *
 * Return value : 
 *      pointer to the data of the removed node, if succeded.
 *      NULL if failed.  
*/
void *SortPopBack(sorted_list_t *sorted_list);


/* Apply a provided function on a given sequence of elements. 
 *
 * NOTE: IF DATA IS ALTERED WITH <action function>, SORTING IS NOT GUARENTEED.
 *
 * Params: 
 *      from - start position.
 *      to - end position (up to and not including).
 *      func - user provided function.
 *      param - additional parameter provided by user.
 *
 * Return value : 
 *      status - success(0) or failure(1). 
*/
int SortForEach(sorted_iter_t from, sorted_iter_t to,
						 	          sort_action_function_t func, void *param);


/* Search if a given data exists on a provided sorted list. 
 *
 * Params: 
 *      sorted_list - a sorted list from which the sequence is given.
 *      from - start position.
 *      to - end position (up to and not including).
 *      param - data to find.
 *
 * Return value : 
 *      pointer to the match node,
 *      or tail dummy if no match.
*/
sorted_iter_t SortFind(sorted_list_t *sorted_list, sorted_iter_t from, 
                  sorted_iter_t to, sort_is_match_function_t func, void *param);


/* Merge two sorted lists into one sorted list <to>. 
 * <from> will be destroyed.
 *
 * Params: 
 *      to - sorted list to merge <second_list> into
 *      from - sorted list to merge.
 *
 * Time Complexity: O(n).
*/
void SortMerge(sorted_list_t *to, sorted_list_t *from);

#endif /* _SORTED_LIST_ */
