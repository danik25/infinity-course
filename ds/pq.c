/****************************************************************
*                <<<<<<<< ASSIGNMENT NAME >>>>>>>>              *
* Related files: Priority.c, Priority_test.c                    *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Or Stopiky                                          *
* Last update: 17/04/19                                         *
****************************************************************/ 
#include <stdlib.h> /*free*/
#include <assert.h> /*assert*/

#include "sorted_list.h"
#include "pq.h"

struct pqueue {
    
    sorted_list_t *sorted_list;
    
};

pqueue_t *PQCreate(int (is_before)(void *data1, void *data2, void *param),
                                                                  void *param)
{
    pqueue_t *s_q = (pqueue_t *) malloc(sizeof(pqueue_t));
    sorted_list_t *s_list = NULL;

    if(!s_q)
    {
        return (NULL);
    }
    
    s_list = SortCreate(is_before, param);

    if(!s_list)
    {
        free(s_q);
        return (NULL);
    }

    s_q->sorted_list = s_list;
    return (s_q);
}


void PQDestroy(pqueue_t *queue)
{
    assert(queue);

    SortDestroy(queue->sorted_list);
    free(queue);
    queue = NULL;
}

int PQEnqueue(pqueue_t *queue, void *data)
{
    assert(queue);

    if(SortInsert(queue->sorted_list, data))
    {
        return(0);
    }
    return (1);
}

void *PQDequeue(pqueue_t *queue)
{
    assert(queue);

    return (SortPopFront(queue->sorted_list));
}

void *PQPeek(const pqueue_t *queue)
{
    assert(queue);

    if(PQIsEmpty(queue))
    {
        return (NULL);
    }

    return (SortGetData(SortBegin(queue->sorted_list)));
}

int PQIsEmpty(const pqueue_t *queue)
{
    assert(queue);

    return (SortIsEmpty(queue->sorted_list));
}

size_t PQSize(const pqueue_t *queue)
{
    assert(queue);

    return (SortSize(queue->sorted_list));
}

void PQClear(pqueue_t *queue)
{
    assert(queue);

    while(!PQIsEmpty(queue))
    {
        PQDequeue(queue);
    }
}

void *PQRemove(pqueue_t *queue, int(* is_match_func)(void *data, void *arge),
                                                                    void *data)
{
    sorted_iter_t node = NULL;
    void * removed_data = NULL;
    node = SortFind(queue->sorted_list, SortBegin(queue->sorted_list), 
        SortEnd(queue->sorted_list), is_match_func, data);
    if(node != SortEnd(queue->sorted_list))
    {
        removed_data = SortGetData(node);
        SortErase(node);
    }
    return (removed_data);
}