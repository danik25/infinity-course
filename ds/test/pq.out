/home/xyz/git/dani-benjamin/ds/pq.out